plugins/ccgx/fu-ccgx-pure-hid-device.c

```diff
@@ -0,0 +1,538 @@
+	}
+	g_debug("After flashing mode");
+	return TRUE;
+}
+
+static gboolean
+fu_ccgx_pure_hid_magic_unlock(FuDevice *device, gpointer user_data, GError **error)
+{
+	guint8 magic_buf[8] = {CCGX_HID_CUSTOM, 0x42, 0x43, 0x59, 0x00, 0x00, 0x00, 0x0B};
+	gboolean err;
+	g_debug("magic unlock");
+
+	if (!fu_hid_device_set_report(FU_HID_DEVICE(device),
+				      magic_buf[0],
+				      magic_buf,
+				      sizeof(magic_buf),
+				      FU_CCGX_PURE_HID_DEVICE_TIMEOUT,
+				      FU_HID_DEVICE_FLAG_IS_FEATURE,
+				      error)) {
+		g_prefix_error(error, "magic enable command error: ");
+		return FALSE;
+	}
+
+
+	g_debug("bridge mode");
+	err = fu_ccgx_pure_hid_command(device, CCGX_HID_CMD_MODE, CY_PD_BRIDGE_MODE_CMD_SIG, NULL);
+	// Ignoring error. This always fails but has the correct behavior
+	err = err;
+
+	return TRUE;
+}
+
+static gboolean
+fu_ccgx_pure_hid_get_fw_info(FuDevice *device, gpointer user_data, GError **error)
+{
+	FuCcgxPureHidDevice *self = FU_CCGX_PURE_HID_DEVICE(device);
+	gsize bufsz = 0x40;
+	g_autofree guint8 *buf = g_malloc0(bufsz);
+	HidFwInfo *info;
+	g_autofree gchar *bl_ver;
+	g_autofree gchar *ver1;
+	g_autofree gchar *ver2;
+	g_debug("get fw info");
+
+	buf[0] = CCGX_HID_INFO_E0;
+	if (!fu_hid_device_get_report(FU_HID_DEVICE(device),
+				      buf[0],
+				      buf,
+				      bufsz,
+				      FU_CCGX_PURE_HID_DEVICE_TIMEOUT,
+				      FU_HID_DEVICE_FLAG_IS_FEATURE,
+				      error))
+		return FALSE;
+
+	fu_ccgx_pure_hid_flashing_mode(device, user_data, error);
+
+	info = (HidFwInfo *)(void *)buf;
+
+	self->operating_mode = info->operating_mode;
+	self->silicon_id = info->silicon_id;
+	self->versions[FU_CCGX_FW_MODE_BOOT] = info->bl_version;
+	self->versions[FU_CCGX_FW_MODE_FW1] = info->image_1_ver;
+	self->versions[FU_CCGX_FW_MODE_FW2] = info->image_2_ver;
+	g_debug("Report ID:       0x%02x", info->report_id);
+	g_debug("Signature:       0x%02x", info->signature);
+	g_debug("Operating Mode:  %s (0x%02x)", fu_ccgx_fw_mode_to_string(info->operating_mode), info->operating_mode);
+	g_debug("Bootloader Info: 0x%02x", info->bootloader_info);
+	g_debug("Bootmode Reason: 0x%02x", info->bootmode_reason);
+	g_debug("Silicon ID:      0x%08x", info->silicon_id);
+	bl_ver = fu_version_from_uint32(info->bl_version, fu_device_get_version_format(self));
+	g_debug("BL Version:      %s (0x%08x)", bl_ver, info->bl_version);
+	ver1 = fu_version_from_uint32(info->image_1_ver, fu_device_get_version_format(self));
+	g_debug("Image 1 Ver:     %s (0x%08x)", ver1, info->image_1_ver);
+	ver2 = fu_version_from_uint32(info->image_2_ver, fu_device_get_version_format(self));
+	g_debug("Image 2 Ver:     %s (0x%08x)", ver2, info->image_2_ver);
+	g_debug("Image 1 Row:     0x%08x", info->image_1_row);
+	g_debug("Image 2 Row:     0x%08x", info->image_2_row);
+	g_debug("Device UID:      %02X%02X%02X%02X%02X%02X", info->device_uid[0], info->device_uid[1], info->device_uid[2], info->device_uid[3], info->device_uid[4], info->device_uid[5]);
```

Comment:
this is begging to be moved into a `to_string` function

Reply:
That's awesome!